File: mediasfu_flutter/lib/room_backend.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:mediasfu_sdk/mediasfu_sdk.dart';

const _backendBaseUrl = 'https://your-app.example/api/mediasfu';

Future<CreateJoinRoomResult> _postRoom(String operation, Map<String, dynamic> payload) async {
  final base = Uri.parse(_backendBaseUrl);
  if (base.scheme != 'https' || base.userInfo.isNotEmpty || base.query.isNotEmpty || base.fragment.isNotEmpty) {
    throw StateError('Configure a plain HTTPS backend base URL.');
  }
  final response = await http.post(
    base.resolve('${base.path.endsWith('/') ? '' : '/'}rooms/$operation'),
    headers: const {'Content-Type': 'application/json'},
    body: jsonEncode(payload),
  );
  final value = jsonDecode(response.body) as Map<String, dynamic>;
  if (response.statusCode < 200 || response.statusCode >= 300) {
    return CreateJoinRoomResult(data: CreateJoinRoomError.fromJson(value), success: false);
  }
  return CreateJoinRoomResult(data: CreateJoinRoomResponse.fromJson(value), success: true);
}

Future<CreateJoinRoomResult> createRoomViaBackend(CreateMediaSFUOptions options) =>
    _postRoom('create', options.payload.toMap());

Future<CreateJoinRoomResult> joinRoomViaBackend(JoinMediaSFUOptions options) =>
    _postRoom('join', options.payload.toMap());

File: mediasfu_react_native_expo/roomBackend.ts
type RoomOperationOptions = { payload: unknown };

const BACKEND_BASE_URL = 'https://your-app.example/api/mediasfu';

async function postRoom(path: 'create' | 'join', payload: unknown) {
  const base = new URL(BACKEND_BASE_URL);
  if (base.protocol !== 'https:' || base.username || base.password || base.search || base.hash) {
    throw new Error('Configure a plain HTTPS backend base URL.');
  }
  const response = await fetch(`${base.toString().replace(/\/$/, '')}/rooms/${path}`, {
    method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify(payload),
  });
  const result = await response.json();
  if (!response.ok) throw new Error(`Room request failed (${response.status})`);
  return result;
}

export const createRoomViaBackend = ({ payload }: RoomOperationOptions) => postRoom('create', payload);
export const joinRoomViaBackend = ({ payload }: RoomOperationOptions) => postRoom('join', payload);
